dynamiclinks/DynamicLinksSwiftUIExample/LinkCreatorExample.swift

```diff
@@ -0,0 +1,133 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+import FirebaseDynamicLinks
+
+struct LinkCreatorExample {
+  enum LinkGenerationError: Error {
+    case missingLinkTarget
+    case missingDomainURIPrefix
+    case invalidLinkPayload
+    case invalidLinkParameters
+    case linkGenerationFailed
+  }
+
+  let parameterStates: [LinkParameter.ID: LinkParameterState]
+
+  func generateDynamicLinkComponents() throws -> DynamicLinkComponents {
+    // general link params
+    guard let linkURL = URL(string: parameterStates[LinkParameter.link.id]?.value ?? "") else {
+      throw LinkGenerationError.missingLinkTarget
+    }
+
+    guard let domainURIPrefix: String = parameterStates[LinkParameter.domainURIPrefix.id]?.value
+    else {
+      throw LinkGenerationError.missingDomainURIPrefix
+    }
+
+    guard let components = DynamicLinkComponents(
+      link: linkURL,
+      domainURIPrefix: domainURIPrefix
+    ) else { throw LinkGenerationError.invalidLinkParameters }
+
+    components.analyticsParameters = generateAnalyticsParameters()
+    components.iOSParameters = generateiOSParameters()
+    components.iTunesConnectParameters = generateiTunesConnectParameters()
+    components.androidParameters = generateAndroidParameters()
+    components.socialMetaTagParameters = generateSocialMetaTagParameters()
+    components.otherPlatformParameters = generateOtherPlatformParameters()
+
+    guard components.url != nil else { throw LinkGenerationError.linkGenerationFailed }
+
+    return components
+  }
+
+  func generateAnalyticsParameters() -> DynamicLinkGoogleAnalyticsParameters? {
+    guard let source = parameterStates[LinkParameter.source.id]?.value,
+      let medium = parameterStates[LinkParameter.medium.id]?.value,
+      let campaign = parameterStates[LinkParameter.campaign.id]?.value else { return nil }
+
+    let analyticsParams = DynamicLinkGoogleAnalyticsParameters(
+      source: source,
+      medium: medium,
+      campaign: campaign
+    )
+    analyticsParams.term = parameterStates[LinkParameter.term.id]?.value
+    analyticsParams.content = parameterStates[LinkParameter.content.id]?.value
+
+    return analyticsParams
+  }
+
+  func generateiOSParameters() -> DynamicLinkIOSParameters? {
+    guard let bundleID = parameterStates[LinkParameter.bundleID.id]?.value else {
+      return nil
+    }
+
+    let iOSParams = DynamicLinkIOSParameters(bundleID: bundleID)
+    iOSParams.fallbackURL = parameterStates[LinkParameter.iOSFallbackURL.id]
+      .flatMap { URL(string: $0.value) }
```

Comment:
Thanks for the suggestion and the article. I skipped this one since I'm no longer using the closure here but I took this approach in the rest of the quickstart.